compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1271,6 +1279,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err.emit();
     }
 
+    fn suggest_ptr_null_mut(
+        &self,
+        expected_ty: Ty<'tcx>,
+        provided_ty: Ty<'tcx>,
+        arg: &hir::Expr<'tcx>,
+        err: &mut rustc_errors::DiagnosticBuilder<'tcx, ErrorGuaranteed>,
+    ) {
+        if let ty::RawPtr(ty::TypeAndMut { mutbl: expected_mutbl, .. }) = expected_ty.kind()
+            && let ty::RawPtr(ty::TypeAndMut { mutbl: provided_mutbl, .. }) = provided_ty.kind()
+            && let hir::ExprKind::Call(callee, _) = arg.kind
+            && let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = callee.kind
+            && let Res::Def(_, def_id) = path.res
+            && let Some(ptr_null_def_id) = self.tcx.get_diagnostic_item(sym::ptr_null)
+            && def_id == ptr_null_def_id
+            && *expected_mutbl == hir::Mutability::Mut
+            && *provided_mutbl == hir::Mutability::Not
```

Comment:
I'm not sure why you don't want to just `ty::RawPtr(ty::TypeAndMut { mutbl: hir::Mutability::Mut, .. })` in the pattern above, as imo it reads quite good: `expected_ty is a raw pointer with mut mutability`...

But if you don't like to do it in the pattern, at least use `.is_mut()`/`.is_not()` and move those two checks to the top, just after the corresponding `let raw_ptr = ty.kind()`-ish code.